### PR TITLE
Add support for creating view with properties in engine and hive

### PR DESCRIPTION
**Pull request by Praveen2112**
## Description

This PR adds support for `CREATE VIEW WITH PROPERTIES` in engine and in Hive connector. In case of hive we are introducing a new property `extra_property` which would allow us to add arbitrary property to hive as `Map<String, String>. 

This PR doesn't have docs and it will be handled as dedicated task which is mentioned as a part of this ticket - https://github.com/trinodb/trino/issues/20992 

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
(x) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Hive
* Add support creating view with properties . ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java [hunk 1]
```diff
@@ -47,15 +48,15 @@
...
 {
-    private final Metadata metadata;
+    private final PlannerContext plannerContext;
     private final AccessControl accessControl;
     private final SqlParser sqlParser;
     private final AnalyzerFactory analyzerFactory;
 
     @Inject
-    public CreateViewTask(Metadata metadata, AccessControl accessControl, SqlParser sqlParser, AnalyzerFactory analyzerFactory)
+    public CreateViewTask(PlannerContext plannerContext, AccessControl accessControl, SqlParser sqlParser, AnalyzerFactory analyzerFactory)
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java]: > Minor cleanup in CreateViewTask

-> `Pass PlannerContext to CreateViewTask`

File: core/trino-main/src/test/java/io/trino/connector/MockConnector.java [hunk 2]
```diff
@@ -768,6 +777,16 @@ public Optional<ConnectorViewDefinition> getView(ConnectorSession session, Schem
             return Optional.ofNullable(getViews.apply(session, viewName.toSchemaTablePrefix()).get(viewName));
         }
 
+        @Override
+        public Map<String, Object> getViewProperties(ConnectorSession session, SchemaTableName viewName)
+        {
+            ImmutableMap.Builder<String, Object> viewProperties = ImmutableMap.builder();
+            for (PropertyMetadata<?> propertyMetadata : getViewProperties.get()) {
+                viewProperties.put(propertyMetadata.getName(), propertyMetadata.getDefaultValue());
```
Comment (findepi) [inline core/trino-main/src/test/java/io/trino/connector/MockConnector.java:785]: or `Maps.uniqueIndex`, or `collect(toImmutableMap())`

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java [hunk 3]
```diff
@@ -972,6 +972,16 @@ default void finishMerge(ConnectorSession session, ConnectorMergeTableHandle mer
...
+    default void createView(ConnectorSession session, SchemaTableName viewName, ConnectorViewDefinition definition, Map<String, Object> viewProperties, boolean replace)
+    {
+        if (viewProperties.isEmpty()) {
+            createView(session, viewName, definition, replace);
+            return;
+        }
+        throw new TrinoException(NOT_SUPPORTED, "This connector does not support creating views");
+    }
+
+    @Deprecated
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:984]: Add javadoc with `@deprecated` instructing what to use instead.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 4]
```diff
@@ -2672,19 +2673,36 @@ private static boolean exists(TrinoFileSystem fs, Location location)
     }
 
     @Override
-    public void createView(ConnectorSession session, SchemaTableName viewName, ConnectorViewDefinition definition, boolean replace)
+    public void createView(ConnectorSession session, SchemaTableName viewName, ConnectorViewDefinition definition, Map<String, Object> viewProperties, boolean replace)
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:2676]: wrap?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 5]
```diff
@@ -2763,6 +2781,13 @@ public List<SchemaTableName> listViews(ConnectorSession session, Optional<String
                 .collect(toImmutableList());
     }
 
+    @Override
+    public Map<String, Object> getViewProperties(ConnectorSession session, SchemaTableName viewName)
+    {
+        // Extra Properties are not exposed via SHOW CREATE VIEW type of queries
+        return ImmutableMap.of();
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:2788]:
```
// The only currently existing view property (extra_properties) is hidden,
// no need to retrieve its value for SHOW CREATE VIEW
```

Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:2788]: How do we know `getViewProperties` is used for SHOW CREATE VIEW only / used only on context where hidden properties do not matter?

Maybe the method name could be `getVisibleViewProperties`?

**Review by findepi — APPROVED** [review]
Syntax already approved in https://github.com/trinodb/trino/pull/21205

File: core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java (hunk 1 above)
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java]: (is this change needed?)

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java (hunk 4 above)
Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:2676]: Currently it is wrapped in a single line right ? Or do we need to move each arguments to a new line ?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java (hunk 5 above)
Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java]: Like `MaterializedView#getMaterializedViewProperties` this is used only for show queries. - `PropertyMetadata#isHidden` is not currently used in show queries. We could rename it to `getVisibleViewProperties` but remaining methods like the one used for MV doesn't fit that pattern so do we need to modify them as well ?

**Comment by Praveen2112** [comment]
@findepi Thanks for the review. AC

File: core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java (hunk 1 above)
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java]: so the `PlannerContext` is indeed used in the second commit? then it's fine

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java (hunk 4 above)
Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:2676]: i meant the line is long, splitting across lines would be appropriate

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java (hunk 5 above)
Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java]: Add `@Experimental` and merge as is. 

let's rename in a follow-up.

File: core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java [hunk 6]
```diff
@@ -74,9 +87,9 @@ public ListenableFuture<Void> execute(
             List<Expression> parameters,
             WarningCollector warningCollector)
     {
-        if (!statement.getProperties().isEmpty()) {
-            throw semanticException(NOT_SUPPORTED, statement, "Creating views with properties is not supported");
-        }
+        Metadata metadata = plannerContext.getMetadata();
+        Map<NodeRef<Parameter>, Expression> parameterLookup = bindParameters(statement, parameters);
```
Comment (marcinsbd) [inline core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java:91]: the bindParameters function seems to be called twice within this method (line 91 , line 112)with same arguments, is it feasible to get rid the second call and reuse result from the first call?

**Comment by Praveen2112** [comment]
@marcinsbd AC

**Review by marcinsbd — APPROVED** [review]

**Pull request merged**